### PR TITLE
[proxy] Symlink sockets to /var/run/weave/weave.sock from inside the proxy

### LINKED ISSUE #1694: Automatically symlink when proxying unix domain sockets under `/var/run`
The proxy currently generates a warning when you attempt to proxy sockets whose immediate parent is `/var/run`, due to the issue with mount namespaces capturing network namespaces. Users are advised to make a symlink of the target socket in a subdirectory and supply that instead; this advice should be automated.

### PR DESCRIPTION
**Pull request by paulbellamy**
Needs an in-depth review, please.

We copy docker.sock owner to symlinks. Permissions are not copied,
because the permissions will inherit from the target.

Fixes #1694, using option 2 (having the proxy create the symlinks). If the weave script creates the symlinks, `weave env` would break (unless we passed the symlinks in, and that's awkward).

There is a slight change in behaviour of `weave env`. New order of preference is: tcp sockets, custom unix sockets, then `/var/run/weave/weave.sock`.

Most of the awkwardness/code in this is just to run a docker container. :disappointed:

Happy to refactor, any suggestions on ways to clean up the code would be great.

### CODE REVIEW

File: proxy/proxy.go [hunk 1]
```diff
@@ -186,7 +192,12 @@ func (proxy *Proxy) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 func (proxy *Proxy) Listen() []net.Listener {
 	listeners := []net.Listener{}
 	proxy.normalisedAddrs = []string{}
+	unixAddrs := []string{}
 	for _, addr := range proxy.ListenAddrs {
+		if strings.HasPrefix(addr, "unix://") {
```
Comment (awh) [inline proxy/proxy.go]: `proxy.listen` interprets any `protoAndAddr` string that begins with a `/` as a unix domain socket - should we not trap that here too?

File: prog/weaveexec/symlink [hunk 2]
```diff
@@ -0,0 +1,18 @@
+#!/bin/sh
+# This lets us force-symlink a bunch of stuff all at once, so we can minimize
+# running containers.
+
+set -e
+
+[ -z "$DEBUG" ] || set -x
```
Comment (awh) [inline prog/weaveexec/symlink]: Is this deliberately not `WEAVE_DEBUG`?

File: proxy/proxy.go [hunk 3]
```diff
@@ -527,3 +552,74 @@ func (proxy *Proxy) updateContainerNetworkSettings(container jsonObject) error {
...
+		},
+		HostConfig: &docker.HostConfig{Binds: binds},
+	})
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		err2 := proxy.client.RemoveContainer(docker.RemoveContainerOptions{ID: container.ID})
+		if err == nil {
```
Comment (awh) [inline proxy/proxy.go]: I'm not sure this will have the effect you intend, as you are creating shadows of `err` with `:=` later on

Comment (paulbellamy) [inline proxy/proxy.go]: it *does* work, but is a bit subtle. Will try to clean it up a bit.

**Comment by bboreham** [comment]
Does this fix #1579 ?

**Comment by paulbellamy** [comment]
I believe so, yes.

**Pull request merged**